internal/services/communication/email_service_domain_sender_username_resource.go

```diff
@@ -0,0 +1,225 @@
+
+			existing, err := client.Get(ctx, id)
+			if err != nil && !response.WasNotFound(existing.HttpResponse) {
+				return fmt.Errorf("checking for presence of existing %s: %+v", id, err)
+			}
+			if !response.WasNotFound(existing.HttpResponse) {
+				return metadata.ResourceRequiresImport(r.ResourceType(), id)
+			}
+
+			parameters := senderusernames.SenderUsernameResource{
+				Properties: &senderusernames.SenderUsernameProperties{
+					Username: model.Username,
+				},
+			}
+
+			if v := model.DisplayName; v != "" {
+				parameters.Properties.DisplayName = pointer.To(v)
+			}
+
+			if _, err := client.CreateOrUpdate(ctx, id, parameters); err != nil {
+				return fmt.Errorf("creating %s: %+v", id, err)
+			}
+
+			metadata.SetID(id)
+
+			return nil
+		},
+	}
+}
+
+func (r EmailCommunicationServiceDomainSenderUsernameResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Communication.SenderUsernamesClient
+
+			var model EmailCommunicationServiceDomainSenderUsernameModel
+
+			if err := metadata.Decode(&model); err != nil {
+				return err
+			}
+
+			id, err := senderusernames.ParseSenderUsernameID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			existing, err := client.Get(ctx, *id)
+			if err != nil || existing.Model == nil {
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			senderUsername := *existing.Model
+
+			props := pointer.From(senderUsername.Properties)
+
+			if metadata.ResourceData.HasChange("username") {
+				props.Username = model.Username
+			}
+
+			if metadata.ResourceData.HasChange("display_name") {
+				props.DisplayName = pointer.To(model.DisplayName)
+			}
+
+			senderUsername.Properties = &props
+
+			if _, err := client.CreateOrUpdate(ctx, *id, senderUsername); err != nil {
+				return fmt.Errorf("updating %s: %+v", id, err)
+			}
+
+			return nil
+		},
+	}
+}
+
+func (EmailCommunicationServiceDomainSenderUsernameResource) Read() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Communication.SenderUsernamesClient
+
+			state := EmailCommunicationServiceDomainSenderUsernameModel{}
+
+			id, err := senderusernames.ParseSenderUsernameID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			resp, err := client.Get(ctx, *id)
+			if err != nil {
+				if response.WasNotFound(resp.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+
+				return fmt.Errorf("retrieving %s: %+v", id, err)
+			}
+
+			state.Name = id.SenderUsernameName
+			state.EMailServiceDomainID = senderusernames.NewDomainID(id.SubscriptionId, id.ResourceGroupName, id.EmailServiceName, id.DomainName).ID()
```

Comment:
could you put `state` initialization here?

Reply:
Updated